data/transactions/verify/txn.go

```diff
@@ -314,3 +329,116 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
 	return nil
 
 }
+
+// PaysetGroups verifies that the payset have a good signature and that the underlying
+// transactions are properly constructed.
+// Note that this does not check whether a payset is valid against the ledger:
+// a PaysetGroups may be well-formed, but a payset might contain an overspend.
+//
+// This version of verify is performing the verification over the provided execution pool.
+func PaysetGroups(ctx context.Context, payset [][]transactions.SignedTxn, blk bookkeeping.Block, verificationPool execpool.BacklogPool) (err error) {
+	txnGroupIdx := 0
+	proto, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]
+	if !ok {
+		return protocol.Error(blk.BlockHeader.CurrentProtocol)
+	}
+	if len(payset) == 0 {
+		return nil
+	}
+	spec := transactions.SpecialAddresses{
+		FeeSink:     blk.BlockHeader.FeeSink,
+		RewardsPool: blk.BlockHeader.RewardsPool,
+	}
+	var txnGroups [][]transactions.SignedTxn
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, concurrentWorksets)
+	worksDoneCh := make(chan interface{}, concurrentWorksets)
+	zeroAddress := basics.Address{}
+	processing := 0
+	tasksCtx, cancelTasksCtx := context.WithCancel(ctx)
+	defer cancelTasksCtx()
+	for processing >= 0 {
+		// see if we need to get another workset
+		if txnGroupIdx < len(payset) && (len(txnGroups) == 0) {
+			txnCounter := 0 // how many transaction we already included in the current workset.
+			// scan starting from the current position until we filled up the workset.
+			for i := txnGroupIdx; i < len(payset); i++ {
+				if txnCounter+len(payset[i]) > txnPerWorksetThreshold {
+					if i == txnGroupIdx {
+						i++
+					}
+					txnGroups = payset[txnGroupIdx:i]
+					txnGroupIdx = i
+					break
+				}
+				if i == len(payset)-1 {
+					txnGroups = payset[txnGroupIdx:]
+					txnGroupIdx = len(payset)
+					break
+				}
+				txnCounter += len(payset[i])
+			}
+		}
+
+		select {
+		case <-tasksCtx.Done():
+			return tasksCtx.Err()
+		case worksets <- struct{}{}:
+			if len(txnGroups) > 0 {
+				err := verificationPool.EnqueueBacklog(ctx, func(arg interface{}) interface{} {
+					// check if we've canceled the request while this was in the queue.
+					if tasksCtx.Err() != nil {
+						return tasksCtx.Err()
+					}
+					txnGroups := arg.([][]transactions.SignedTxn)
```

Comment:
I still think `verify.Txn` should be called here inside loops. most likely the benchmark will not show any noticeable difference but the code would be clearer